### PR TITLE
fix: Allow notifications controller to notify on all namespaces

### LINKED ISSUE #11252: Notifications for applications in namespaces do not work
**Describe the bug**

We migrated our applications to specific namespaces which made it easier for us to logically group applications within Kubernetes. After we migrated them we also wanted to configure notifications for them but it does not work. We had a quick look into it and it seems that first, that the notification controller does not have permissions to get all applications in all namespaces (a ClusterRole is missing). Second, we are not totally sure, if the controller itself can manage applications within other namespaces (we were not able to figure it out with a quick look over the code - but we didn't spent too much time looking for it).

In this specific case we wanted to receive Slack notifications. But I don't think that it would work with other services either. We were not able to see any activity of the application controller whatsoever.

For test purposes we configured an application in the argocd namespace. When we configured a notification subscription for it it immidiatly worked and we were able to see logs in the notification controller.

**To Reproduce**

- Deploy at least ArgoCD 2.5.0
- Allow ArgoCD to manage applications in other namespaces
- Configure notification templates and triggers
- Deploy an application into an allowed namespace (but not `argocd`)
- Configure notficiation subscription (either global or application annotation)

**Expected behavior**

The notification controller can monitor all applications in the cluster and react to events.

**Version**

```shell
argocd: v2.5.2+148d8da
  BuildDate: 2022-11-07T16:42:47Z
  GitCommit: 148d8da7a996f6c9f4d102fdd8e688c2ff3fd8c7
  GitTreeState: clean
  GoVersion: go1.18.8
  Compiler: gc
  Platform: linux/amd64
```

### LINKED ISSUE #12912: AgroCD Notifications doesn't work with different namespaces
Hello team,
I have been trying to implement ArgoCd notifications into our clusters. First let me explain my use case and scenario.
We have k8s clusters running and they are provisioned and maintained my a centralised team. They have installed ArgoCd and also notification in their foundation namespace which we use with our applications and configs. We do not have control on their namespaces so I installed ArgoCD notification controller separately in a different namespace.
Will all configured I can see the notification pod logs as below -
`time="2023-02-23T16:02:47Z" level=info msg="ArgoCD Notifications Controller is starting" built="2023-02-16T15:05:14Z" commit=6e02f8b23201b0620a4ff1bce5d38229ba1eb02e namespace=infra version=v2.6.2+6e02f8b
3
{"level":"info","msg":"serving metrics on port 9001","time":"2023-02-23T16:02:47Z"}
2
{"level":"info","msg":"loading configuration 9001","time":"2023-02-23T16:02:47Z"}
1
{"level":"warning","msg":"Controller is running.","time":"2023-02-23T16:02:47Z"}
`

When I try to delete a deployment I get no teams message and no logs to troubleshoot. Am I doing something wrong?

This is the deployment image I use "**_quay.io/argoproj/argocd:v2.6.2_**"

The cm is configured as below and the templates and triggers are copied from the latest documentation

```
apiVersion: v1
  data:
    context: |
      argocdUrl: $argourl
      environmentName: dev
      region: west-europe
    service.teams: |
      recipientUrls:
        channelName: $msteams
    service.webhook.msteams: |
      url: $msteams
    subscriptions: |
      - recipients:
        - msteams
        triggers:
        - on-sync-succeeded
        - on-created
        - on-deleted
        - on-deployed
        - on-health-degraded
        - on-sync-failed
        - on-sync-running
        - on-sync-status-unknown
```

### PR DESCRIPTION
**Pull request by Niksko**
This adds functionality to the notifications controller to be notified of and send notifications for applications in any namespace. The namespaces to watch are controlled by the same --application-namespaces and ARGOCD_APPLICATION_NAMESPACES variables as in the application controller.

Closes #12912
Closes #11252 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] TODO: I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

**Comment by Niksko** [comment]
Hi folks, I'd appreciate a few pointers:

* Is there documentation for flags passed to the various commands, other than in the commands themselves? If so can you please point me to them, I'd be happy to update them.
* What sorts of tests are appropriate for this PR? I can't find any existing tests.

**Comment by iandelahorne** [comment]
I'm facing this exact problem - we use Applications In Any Namespace, and would like to onboard Argo notifications. Thanks a lot for implementing this 🙏  

I tried this out in my cluster and found that I needed to add a ClusterRole with permissions necessary and bind it to the notifications controller's service account. 

Adding a clusterrole and clusterrolebinding to the manifests/cluster-rbac/notifications directory would make this smoother. Also, the deployment should pull `ARGOCD_APPLICATION_NAMESPACES` from the `argocd-cmd-params-cm` configmap similar to the applicationset controller does [here](https://github.com/argoproj/argo-cd/blob/7825821c1cda17b809e44b66cc5980fecf28af4f/manifests/base/applicationset-controller/argocd-applicationset-controller-deployment.yaml#L40-L45)

**Comment by iandelahorne** [comment]
Another thing that came to mind is the templates in the notifications catalog use `oncePer: app.metadata.name`([example](https://github.com/argoproj/argo-cd/blob/master/notifications_catalog/triggers/on-created.yaml#L4)) which we would need to handle somehow.

**Comment by Niksko** [comment]
@iandelahorne thanks for the feedback, I'll add in those permissions and handling of the once per namespace template field

**Comment by jaideepr97** [comment]
@Niksko 

https://argo-cd.readthedocs.io/en/stable/operator-manual/app-any-namespace/#change-workload-startup-parameters

The app namespace list can also be set in the `argocd-cmd-params-cm`, which I don't think the notifications-controller tracks today
https://github.com/argoproj/argo-cd/blob/master/manifests/install.yaml#L18874

Maybe that could also be addressed in this PR?

**Comment by major7x** [comment]
Hallo! any updates on this PR? Would be greatly appreciated if this is looked at as it would be greatly beneficial. Thanks

File: notification_controller/controller/controller.go [hunk 1]
```diff
@@ -53,14 +54,15 @@ func NewController(
...
 	appLabelSelector string,
 	registry *controller.MetricsRegistry,
 	secretName string,
 	configMapName string,
 ) *notificationController {
 	appClient := client.Resource(applications)
-	appInformer := newInformer(appClient.Namespace(namespace), appLabelSelector)
-	appProjInformer := newInformer(newAppProjClient(client, namespace), "")
+	appInformer := newInformer(appClient, namespace, applicationNamespaces, appLabelSelector)
+	appProjInformer := newInformer(newAppProjClient(client), namespace, applicationNamespaces, "")
```
Comment (crenshaw-dev) [inline notification_controller/controller/controller.go:65]: Projects still just live in the `argocd` namespace, so this informer shouldn't look at all `applicationNamespaces`.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

**Comment by Niksko** [comment]
Apologies folks, I do intend to get back to this, ideally within the next few weeks. If someone else wants to pick this up in the mean time and submit a fix I won't be offended, I acknowledge that I've let this drag on

**Comment by sthomson-wyn** [comment]
Addressed comments from @iandelahorne @jaideepr97 @crenshaw-dev here: https://github.com/argoproj/argo-cd/pull/15247

**Comment by major7x** [comment]
Thank you @sthomson-wyn! @crenshaw-dev would it suffice to proceed?

**Comment by crenshaw-dev** [comment]
Superseded by https://github.com/argoproj/argo-cd/pull/15702

**Pull request closed without merging**